src/python/WMCore/MicroService/Unified/MSMonitor.py

```diff
@@ -17,123 +21,170 @@ class MSMonitor(MSCore):
                 # if pileup transfers are completed AND
                 # some input blocks are completed,
                 # move the request status staging -> staged
-                elif self.pileupTransfersCompleted(transferRecords):
+                elif self.completion(transfers, pileup=True):
                     self.logger.debug(
-                        "+++ request %s pileup transfers are completed", req)
-                    self.change(req, 'staged', '+++ monit')
+                        "+++ request %s pileup transfers are completed", reqName)
+                    completedWorkflows.append(wname)
+                    reqStatus[reqName] = 'staged'
                 # transfers not completed
                 # just update the database with their completion
                 else:
                     self.logger.debug(
-                        "+++ request %s transfers are %s completed",
-                        req, completion)
-                    # for us workflow name is the same as request name
-                    wname = reqName
-                    requestStatuses[wname] = transferRecords
-            self.updateTransferInfo(requestStatuses)
+                        "+++ request %s transfers are completed", reqName)
+            self.updateTransferInfo(transferRecords, completedWorkflows)
+            # finally, update statuses for requests
+            for reqName, status in reqStatus.items():
+                self.change(reqName, status, '+++ monit')
         except Exception as err:  # general error
             self.logger.exception('+++ monit error: %s', str(err))
 
-    def getTransferRecords(self, reqName):
+    def completion(self, transfers, pileup=None):
+        """
+        Helper function to calculate completion of given records
+        :param transfers: list of transfers records
+        :param pileup: check pileup completion (optional)
+        :return: completion status
+        """
+        # check completion of all transfers
+        statuses = []
+        transferTypes = ['primary', 'secondary']
+        for rec in transfers:
+            campaign = rec['campaignName']
+            cdict = self.campaigns[campaign]
+            thr = cdict.get('PartialCopy', 0)
+            if pileup and rec['dataType'] not in transferTypes:
+                continue
+            if rec['completion'] >= thr:
+                status = 1
+            else:
+                status = 0
+            statuses.append(status)
+        return True if sum(statuses) == len(transfers) else False
+
+    def updateTransferInfo(self, transferRecords, completedWorkflows):
+        """
+        Update transfer records in backend
+        :param transferRecords: list of transfer records
+        :param completedWorkflows: list of completed workflow names to be removed from backend
+        """
+        tstamp = time.time()
+        for doc in transferRecords:
+            if doc['workflowName'] in completedWorkflows:
```

Comment:
I think we should still update these transfer records, such that we can go back and see what was the last snapshot MSMonitor had for each of those transfers/workflows.

Reply:
and I think they should be discarded, since if you update them their update timestamp will advance. Then we'll need to introduce another thread which will monitor requests statuses to decide when we should wiped out records from CouchDB instead of just relying on timestamp. Do you really want to add another complexity?

Reply:
document cleanup will be based on request status only, no need to make it complex. So updating the timestamp or not makes no difference.